File: 2020/python/day05.py
#!python3

'''
FBFBBFFRLR: row 44, column 5, seat ID 357.
BFFFBBFRRR: row 70, column 7, seat ID 567.
FFFBBBFRRR: row 14, column 7, seat ID 119.
BBFFBBFRLL: row 102, column 4, seat ID 820.
'''


INPUT = 'day05.txt'

TEST_INPUT = '''FBFBBFFRLR
BFFFBBFRRR
FFFBBBFRRR
BBFFBBFRLL'''

TEST_OUTPUT = [
    357, 567, 119, 820
]


TRANSLATION = str.maketrans('FBRL', '0110')

def get_seats(input):
    for row in input.strip().split('\n'):
        yield row[:7], row[7:]


def translate_binary(binary):
    b = binary.translate(TRANSLATION)
    return int(b, 2)


def get_seat_id(seat):
    row, col = map(translate_binary, seat)
    return 8 * row + col


def test1():
    seats = get_seats(TEST_INPUT)
    for (seat, seat_id) in zip(seats, TEST_OUTPUT):
        assert get_seat_id(seat) == seat_id


def part2(seat_ids):
    min_id = min(seat_ids)
    max_id = max(seat_ids)
    seat_set = set(seat_ids)
    all_seat_set = set(range(min_id, max_id))
    empty_set_set = all_seat_set - seat_set
    return empty_set_set.pop()


def main():
    seats = get_seats(open(INPUT, 'r').read())
    seat_ids = [get_seat_id(s) for s in seats]

    test1()

    print(max(seat_ids))

    p = part2(seat_ids)
    print(p)


if __name__ == "__main__":
    main()

File: 2020/c++/day05.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

#include "basepath.cpp"
#include "inputhelpers.cpp"

typedef std::string Pass;
typedef std::vector<std::string> Passes;
typedef int SeatId;

void part1();
void test1();
Passes get_passes();
const SeatId get_seat_id(Pass);
const int get_row(Pass);
const int get_column(Pass);

const Passes TESTS {
    "FBFBBFFRLR",
    "BFFFBBFRRR",
    "FFFBBBFRRR",
    "BBFFBBFRLL"
};

int main() {
    // test1();
    part1();

    return 0;
}


void part1() {
    Passes passes = get_passes();
    std::vector<SeatId> seat_ids;
    for (Pass pass : passes) {
        SeatId seat_id = get_seat_id(pass);
        seat_ids.push_back(seat_id);
    }
    int max = *std::max_element(seat_ids.begin(), seat_ids.end());
    std::cout << "Part 1 " << max << std::endl;
}


Passes get_passes() {
    return get_lines(base_path + "day05.txt");
}


void test1() {
    for (string t : TESTS) {
        std::cout << t << ' ' << get_seat_id(t) << std::endl;
    }
}


const SeatId get_seat_id(std::string code) {
    const int row = get_row(code);
    const int column = get_column(code);
    return 8 * row + column;
}


const int get_row(std::string code) {
    int row { 0 };
    for (int i = 0; i < 7; i++) {
        const char test = code[i];
        if (test == 'F') {
            row = row << 1;
        } else if (test == 'B') {
            row = (row << 1) + 1;
        } else {
            std::cout << "row char error " << test << std::endl;
        }
    }
    return row;
}


const int get_column(std::string code) {
    int column { 0 };
    for (int i = 7; i < 10; i++) {
        const char test = code[i];
        if (test == 'R') {
            column = (column << 1) + 1;
        } else if (test == 'L') {
            column = column << 1;
        } else {
            std::cout << "column char error " << code[i] << std::endl;
        }
    }
    return column;
}